### PR TITLE
add and remove tags from multiple files at once

### PR DESCRIPTION
**Pull request by rolandinus**
#10674 
Adds an option to the actions when multiple files are selected.
The tags from all selected files are fetched and the tags which are common to all files will be preselected.
When Tags are deselected they will be removed from all selected files.
When new Tags are selected or created they will be added to all files

### CODE REVIEW

**Comment by kesselb** [comment]
Could you rebase the branch? Should fix the test failures.

**Comment by skjnldsv** [comment]
@danxuliu could you give a hand on tests here? :)

**Comment by juliushaertl** [comment]
unit tests should be fixed by rebasing to curent master

**Comment by skjnldsv** [comment]
The rebase doesn't seems right :see_no_evil:

**Comment by rolandinus** [comment]
There was some mess with the rebase, should be ok now

**Comment by skjnldsv** [comment]
ALL GREEN!!!! @nextcloud/javascript can you believe it :D

**Review by skjnldsv — APPROVED** [review]

File: apps/files/js/filelist.js [hunk 1]
```diff
@@ -1070,6 +1073,112 @@
...
+         * @param tagId
+         * @private
+         */
+		_onDeselectTag: function(tagId) {
+            var selectedFiles = _.pluck(this.getSelectedFiles(),'id')
+            console.log('removing tag', tagId, 'for files', selectedFiles, this.getSelectedFiles())
+            if (!_.isArray(selectedFiles)) {
+                console.log('selected files - no array!', selectedFiles);
+            }
+            selectedFiles.forEach(function(fileId) {
```
Comment (ChristophWurst) [inline apps/files/js/filelist.js]: :boom: did you mean to return early two lines above?

**Review by ChristophWurst — CHANGES_REQUESTED** [review]

File: apps/files/js/filelist.js (hunk 1 above)
Comment (rolandinus) [reply apps/files/js/filelist.js]: Yes, that was the plan originally- but I nevery run in to the case while testing that I got not back an array, so I seem to have forgotten it. However if I look at the source getSelectedFiles() it is not guaranteed to return a an array

Comment (ChristophWurst) [reply apps/files/js/filelist.js]: yes, but this is a bit strange. you check the type, but then you continue anyway, using an API of `Array`. You should either assume it's an array (and remove the check) or handle this case properly and either throw an `Error(...)` or abort the current operation.

Comment (rolandinus) [reply apps/files/js/filelist.js]: I changed the code and return early now if no files are selected.

**Comment by skjnldsv** [comment]
Hey @rolandinus 
Now that 18 is released, could you rebase your work to latest master please? :)

**Comment by rolandinus** [comment]
> Hey @rolandinus
> Now that 18 is released, could you rebase your work to latest master please? :)

I rebased it to newest master, but now I get again errors on  some the tests (for server side parts which I did not touch at all) Last time they vanished after rebasing them to the newest master. But I pushed directly after rebasing. Can anyone give me information about why these errors occur and how i can avoid them?

**Comment by skjnldsv** [comment]
> Can anyone give me information about why these errors occur and how i can avoid them?

They're coming from anbother issue introduced in master.
They don't seems to be related to your (awesome) work :)

File: apps/files/js/filelist.js [hunk 2]
```diff
@@ -1080,6 +1083,110 @@
...
+                    multiple: true,
+                    allowActions: true,
+                    allowCreate: true,
+                    isAdmin: OC.isUserAdmin(),
+                });
+                self._inputView.on('select', self._onSelectTag, self);
+                self._inputView.on('deselect', self._onDeselectTag, self);
+                self._inputView.render();
+                self.tagsContainer = $('<tr id="tag_multiple_files_container"><th colspan="4"></th></tr>');
+               // console.log('inputView',self._inputView, self._inputView.$el,'tag container', self.tagsContainer);
```
Comment (kesselb) [inline apps/files/js/filelist.js]:
```suggestion
```

**Comment by TueTchen** [comment]
Dear Experts,
would there be a possibility to try this important function in advance in an existing Nextcloud 18 installation as a beta tester?
We're a small group working honorary on preserving children's rights due to child abuse. We're currently working on a big hard case with over 1500 scanned documents that we've already named so that we can get them filtered through Nextcloud Search. And we have prepared about 50 collaborative tags to get these tons of artefacts correctly prepared for our thesis.
What would help us very very much right now, that would be to have your great development of "mass tagging" to tag about 20 to 300 documents at once.
Would there be any chance to get a pre- or alpha release working in a quite fresh 18.0.4 nextcloud installation on a managed hosting server, without having too deep development knowhow, but having enough to understand most of your discussions?
Thanks in advance for your answer!
Peter alias TueTchen, Germany

**Comment by Spartachetto** [comment]
@kesselb @skjnldsv @juliushaertl I am sorry to bother you, yet this feature would be great... 

Do you think it possible for it to be 
in Nextcloud 19?

Thank you in any case and keep up the good work!!!

**Review by ChristophWurst — COMMENTED** [review]
Code looks okay-ish, while this legacy jquery style is more than just outdated. But that is the state of the files app right now, so I guess this is ok?

Didn't test and don't have the in depth understanding of this part of our code base. Thus there might be uncovered edge cases.

:elephant:

**Comment by rullzer** [comment]
I rebased this for CI

**Comment by jancborchardt** [comment]
> this legacy jquery style is more than just outdated. But that is the state of the files app right now, so I guess this is ok?

Yep – it is indeed outdated but that’s where we are. It’s good for 19 and a big help for people who use tagging.

Short of bigger design feedback I only have 1 point @rolandinus 
- [ ] The "Tags" action should not be the last action in the menu. Better between "Download" and "Select file range". (The last action should always be the destructive action, in this case being "Delete".

File: apps/files/js/app.js [hunk 3]
```diff
@@ -102,6 +102,11 @@
 							displayName: t('files', 'Delete'),
 							iconClass: 'icon-delete',
 						},
+                        {   name: 'tags',
+                            displayName:  'Tags',
+                            iconClass: 'icon-tag'
+
+                        }
```
Comment (jancborchardt) [inline apps/files/js/app.js:109]: This block should go a bit further up so it’s below the "Download" action instead. :) Otherwise all good

**Review by jancborchardt — CHANGES_REQUESTED** [review]
See comment ^

**Review by skjnldsv — CHANGES_REQUESTED** [review]
> Yep – it is indeed outdated but that’s where we are. It’s good for 19 and a big help for people who use tagging.

No, sorry, I have to block.
This is too sensitive. Sorry @rolandinus , this is some areas of the code where it alwyas blow up later on. I refuse to be the one that have to fix 3 legacy files that are unrelated for this to get in.
Beside, 19 is feature freeze, nothing like that gets in for now :)

**Comment by ChristophWurst** [comment]
Those who eagerly want this feature can just apply as as patch. You can download that from https://patch-diff.githubusercontent.com/raw/nextcloud/server/pull/18199.patch.

**Comment by MorrisJobke** [comment]
> No, sorry, I have to block.
> This is too sensitive. Sorry @rolandinus , this is some areas of the code where it alwyas blow up later on. I refuse to be the one that have to fix 3 legacy files that are unrelated for this to get in.
> Beside, 19 is feature freeze, nothing like that gets in for now :)

@rolandinus Unfortunately I have to acknowledge this. We are already in beta 7 and it is a too high risk of breakage here. I will move it to the 20 milestone.

**Comment by rolandinus** [comment]
I know the code is outdated, I just tried to fit it in with the code of the rest of the files app. I am happy to refactor it for a future version of NC. However it would only make sense if the rest of the files app is refactored before I guess

**Comment by skjnldsv** [comment]
@rolandinus Also, I want to give you some explanation, because I think you deserve to have some.
Our files app code is very old. While we're doing a **crazy** amount of work to have a proper implementation of our features in our apps, this is the core of our project and unfortunately it is incredibly messy. This was written in a different time and to be fair, none of us understand the code architecture. This is patched everywhere, contains lots of jQuery code, etc etc.
This makes it very hard to do anything, because any little enhancement always end up in multiple unknown bugs that takes a few days to debug.

It is never worth it.
We have plans to upgrade the files app too, but those things takes time, as youc an imagine. Keep in mind, we really are all in your awesome feature, but I'm too cold feet and cautious to approve this. For me it's not worth the risk (and we can gamble, but it always blow up somewhere unexpected, alway :( )

Hope you can understand why it takes so long for this to get approved.
Thanks again for your work and interest! :hugs: :bowing_woman:

**Comment by TueTchen** [comment]
Thank you all for your efforts and your explanations!
In the meantime I'll help myself/ourselves by using the good old school tool 'nc-direct' (= SQL 😅). The tag relation are really very simple, and our acribic file naming help us to get a main mass tagging done on its lowest level.
Regarding our wider usage of tagging (currently >100 tags for >3000 files), there are also growing new requirement and wishes, that underline a well-thought refactoring - e.g. hierarchical tag relations, access rights enhancements and also another kind of showing and handling file taggings (we've already up to 20 tags for the single file, it's not really easy to identify missing or wrong tags).

So thank you again, always good luck & good speed and please stay healthy!
Peter

**Comment by HTStentiQ** [comment]
any news on this?

**Comment by skjnldsv** [comment]
Closing, see https://github.com/nextcloud/server/pull/18199#issuecomment-623326652

**Comment by Sandro1955** [comment]
Hi, I installed version 12.01, but I can't find the "multiple tags" function in the actions. How do I go about implementing it?

**Comment by Sandro1955** [comment]
Hi, I installed version 21.0.4, but I can't find the "multiple tags" function in the actions. How do I go about implementing it?

**Comment by skjnldsv** [comment]
https://github.com/nextcloud-gmbh/professional_services/issues/321
